File: Python Algorithms/astar.py
import heapq

# Define the heuristic function h
def h(node):
    # Example heuristic: straight-line distance to goal (replace with your actual heuristic)
    return distance[node]

# Define the cost function g
def g(node1, node2):
    # This is typically the actual cost of the edge from node1 to node2
    return graph[node1][node2]

# Define the A* search algorithm
def astar(start, goal):
    queue = []
    heapq.heappush(queue, (0 + h(start), start))
    costs = {start: 0}
    path = {start: [start]}

    while queue:
        (priority, current) = heapq.heappop(queue)
        if current == goal:
            return path[current]

        for next_node in graph[current].keys():
            new_cost = costs[current] + g(current, next_node)
            if next_node not in costs or new_cost < costs[next_node]:
                costs[next_node] = new_cost
                priority = new_cost + h(next_node)
                heapq.heappush(queue, (priority, next_node))
                path[next_node] = path[current] + [next_node]

    return []

# Example data
graph = {
    'a': {'b': 1, 'c': 2},
    'b': {'c': 2},
    'c': {'goal_node': 1},
    'goal_node': {}
}

distance = {
    'a': 3,
    'b': 2,
    'c': 1,
    'goal_node': 0
}

# Usage example
print(astar('a', 'goal_node'))  # Output: ['a', 'b', 'c', 'goal_node']

File: AI_Sem problems/Prolog_Workspace/Astar.pl
% Define the heuristic function h/2.
% This is problem-specific and must be admissible.
h(Node, HeuristicValue) :-
    % Example heuristic: straight-line distance to goal (replace with your actual heuristic).
    goal(Goal),
    distance(Node, Goal, HeuristicValue).

% Define the cost function g/3.
% This is typically the actual cost of the edge from Node1 to Node2.
g(Node1, Node2, Cost) :-
    edge(Node1, Node2, Cost).

% Define the estimated cost function f/3.
f(Node1, Node2, EstimatedCost) :-
    g(Node1, Node2, Cost),
    h(Node2, HeuristicValue),
    EstimatedCost is Cost + HeuristicValue.

% Define the A* search algorithm.
astar(Start, Goal, Path) :-
    astar(Start, Goal, [Start], 0, Path).

astar(Goal, Goal, Path, _, Path).
astar(Current, Goal, Visited, CurrentCost, Path) :-
    findall(
        (Next, NewCost, EstimatedCost, [Next|Visited]),
        (
            edge(Current, Next, StepCost),
            \+ member(Next, Visited),
            NewCost is CurrentCost + StepCost,
            h(Next, HeuristicValue),
            EstimatedCost is NewCost + HeuristicValue
        ),
        Neighbors
    ),
    keysort(Neighbors, SortedNeighbors),
    member((Next, NewCost, _, NewVisited), SortedNeighbors),
    astar(Next, Goal, NewVisited, NewCost, Path).

% Example predicates
% Define goal/1 to specify the goal node
goal(goal_node).

% Define edge/3 to specify the edges and their costs
edge(a, b, 1).
edge(b, c, 2).
edge(a, c, 2).
edge(c, goal_node, 1).

% Define distance/3 to specify heuristic values (e.g., straight-line distances)
distance(a, goal_node, 3).
distance(b, goal_node, 2).
distance(c, goal_node, 1).

% Usage example
% ?- astar(a, goal_node, Path).
% Path = [goal_node, c, a].
